benchmarks/replay/replay.cpp

```diff
@@ -79,6 +88,65 @@ struct replay_benchmark {
   }
 };
 
+/**
+ * @brief Processes a log file into a set of per-thread vectors of events
+ *
+ * @param filename Name of log file
+ * @return A vector of events for each thread in the log
+ */
+std::vector<std::vector<rmm::detail::event>> parse_per_thread_events(std::string const& filename)
+{
+  using rmm::detail::event;
+  std::vector<event> all_events = rmm::detail::parse_csv(filename);
+
+  RMM_EXPECTS(std::all_of(all_events.begin(),
+                          all_events.end(),
+                          [](auto const& e) {
+                            return (e.stream == cudaStreamDefault) or
+                                   (e.stream == reinterpret_cast<uintptr_t>(cudaStreamPerThread));
+                          }),
+              "Non-default streams not currently supported.");
+
+  // Sort events by thread id
+  std::stable_sort(all_events.begin(), all_events.end(), [](auto lhs, auto rhs) {
+    return lhs.thread_id < rhs.thread_id;
+  });
+
+  // Count the number of events per thread
+  std::vector<std::size_t> events_per_thread{};
+  thrust::reduce_by_key(
+    thrust::host,
+    all_events.begin(),
+    all_events.end(),
+    thrust::make_constant_iterator(1),
+    thrust::make_discard_iterator(),
+    std::back_inserter(events_per_thread),
+    [](event const& lhs, event const& rhs) { return lhs.thread_id == rhs.thread_id; });
+
+  auto const num_threads = events_per_thread.size();
+
+  // Per thread offset into the sorted list of events
+  std::vector<std::size_t> thread_event_offsets(num_threads + 1, 0);
+  thrust::inclusive_scan(thrust::host,
+                         events_per_thread.begin(),
+                         events_per_thread.end(),
+                         std::next(thread_event_offsets.begin()));
+
+  // Copy each thread's events into its own vector
+  std::vector<std::vector<event>> per_thread_events(num_threads);
+  thrust::transform(thrust::host,
+                    thread_event_offsets.begin(),
+                    std::prev(thread_event_offsets.end()),
+                    std::next(thread_event_offsets.begin()),
+                    per_thread_events.begin(),
+                    [&all_events](auto begin, auto end) {
+                      return std::vector<event>(all_events.begin() + begin,
+                                                all_events.begin() + end);
+                    });
```

Comment:
Arguably, since this is sequential, it's simpler and briefer to replace the inclusive_scan and the transform with a single sequential transform. Something like this?

```
std::transform(events_per_thread.begin(), events_per_thread.end(), thread_event_offsets.end(), 
               [offset = 0] mutable (auto &num_events) { 
  auto begin = offset;
  offset += num_events;
  auto end = offset;
  return std::vector<event>(all_events.begin() + begin, all_events.begin() + end);
});
```

Reply:
Yeah, that is definitely briefer. Matter of habit to use a scan when confronted with a problem that can be solved by one. 

Guy would be sad :(